src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -1557,6 +1557,22 @@ void MarkField (FieldDefinition field, in DependencyInfo reason)
 				Annotations.Mark (field, reason);
 			}
 
+			switch (reason.Kind) {
+			case DependencyKind.AccessedViaReflection:
```

Comment:
I know we use this kind of pattern in a few places already (I added some of them myself :frowning:) but I think we should avoid it. Special-casing behavior for some "reason kinds" isn't very maintainable - it requires examining all callers of the method to see which cases should be treated specially. I would prefer to add a separate parameter to MarkField that indicates what the optional behavior is, and pass it in from some callers only.

Reply:
In general I agree, but it would require non-trivial effort. Not for fields, but for methods. The entire RUC and DAM detection in `MarkMethod` currently relies on this. If we were to fix it, I think we should do it in both places.

I did go through all the callers for both methods and fields, it's tedious I agree, but fixing it would require lot more work. I didn't want to block this change on it. Mainly because I want to avoid passing many different parameters to Mark methods. And `MarkMethod` would need at least 2 more (RUC and DAM handling). So in my mind a cleaner solution would be to do this in the planned refactoring of scope stacks and `DependencyInfo` into one parameter. With that we could relatively easily add more fields to the structure we will pass around and store additional information that way.